src/nim_status/c/lib/shim.nim

```diff
@@ -0,0 +1,8 @@
+import ../../lib/shim as nim_shim
+import ../sys
+
+proc hashMessage*(message: cstring): cstring =
```

Comment:
I don't quite understand the rationale behind the general strategy here.

The most straigh-forward approach for me for creating such a shim layer would be to define a high-level interface that uses the most natural Nim types for the task at hand. Here for example, you wouldn't return a `cstring`, but rather a normal `string` or even a stronger type such as `nimcrypto.MDigest`.

Once you have this high-level interface defined, you can create a thin translation layer that calls Go or lib_status.a to produce the result with some lower-level pointer-based APIs. The application will be refactored everywhere to make use of the new nicer higher-level API and the long-term prospects are that the lower-level pointer tricks can be removed in favour of directly importing modules from nim-status in the future or by depending on it as a DLL (through the use of [`nimRtl`](https://nim-lang.org/docs/nimc.html#dll-generation)).

I cannot imagine what is gained by imitating the existing API of Status-Go. Since it's too foreign for Nim, it will always have to be wrapped in some facade layer.

Reply:
Thanks for the feedback @zah, much appreciated! I'll try my best to explain, and I'm very open to going a different direction with things.

> Here for example, you wouldn't return a `cstring`, but rather a normal `string` or even a stronger type such as `nimcrypto.MDigest`

For the intended beginnings of that see `src/nim_status/lib.nim`. The idea is to use the most natural Nim types and Nim programming style in `lib`; users of the library are free to `import nim_status/lib` if they don't want to use the shimmed `proc`s that imitate status-go. Granted, I just stuck with normal `string` in `hashMessage` of `src/nim_status/lib.nim`, but as you say it can probably be improved with a richer return type like `nimcrypto.MDigest`.

> Once you have this high-level interface defined, you can create a thin translation layer

That's actually what I was/am going for, but being so new to Nim (as in completely new) I may have gone about it in a way that's not as natural or ergonomic as I could have.

> I cannot imagine what is gained by imitating the existing C API of Status-Go

My intention was/is to do the imitating only in `src/nim_status/go/[c/]/shim.nim`, which is a simple wrapper around status-go, and `src/nim_status/lib/[c/]/shim.nim`, which does the job of imitating status-go for `proc`s in `src/nim_status/lib.nim`. Those two are combined in `src/nim_status.nim` and `src/nim_status/c/nim_status.nim` to provide an imitation that's partly "pure Nim" and partly status-go.

So why have `nim_status.nim` be a combo of shims that imitate the existing C API of status-go?

To my mind, in the beginning steps, as most of the library relies on wrapping status-go, it seems to enable a more direct swap, i.e. a user like status-react could switch from compiling/linking status-go (`libstatus.a`) to compiling/linking nim-status (`nim_status.a`), and it would just work. At the same time, a Nim-based project like `nim-status-client` can use things implemented in Nim via `import nim_status/lib` (with the more natural types) and get things still only available in status-go via `import nim_status/go/shim`.

Theoretically, there comes a point where `src/nim_status/lib.nim` implements *all* of the functionality and the wrapper around status-go becomes redundant. The wrapper (`go/shim.nim`) could be deprecated at that point. If/when a project like status-react is able to completely migrate away from the status-go calling conventions to ones in `src/nim_status/c/lib.nim` then `lib/shim.nim` could be deprecated.

Reply:
I see, I've indeed missed the role of `nim_status/lib`. 

The shims here (in `c/lib/shim`) make sense as long as you have users currently using Status-Go from pure C that are interested in getting a drop-in replacement implemented in Nim (although, you'll have to cater to other Nim-specific details such as ensuring that [`setupForeignThreadGc`](https://nim-lang.github.io/Nim/backends.html) is called from the C code). If you don't have such users though, maintaining that layer can be seen as a currently unnecessary extra work (you can always create it on demand according to the logic behind the [YAGNI principle](https://wiki.c2.com/?YouArentGonnaNeedIt)).

If your goal is to develop nim-desktop-client though, I think a different strategy makes more sense. You can try to create the natural Nim interface as soon as possible and integrate it everywhere where Status-Go is currently used. The initial version will delegate the calls to Status-Go by translating them to the native Nim types and gradually more and more calls will be replaced with native Nim code. At no point, you would have to resort to using `c_malloc` and `copyMem` on the native Nim side.